src/vec.rs

```diff
@@ -0,0 +1,324 @@
+//! ----------------------------------------
+//! E_1                    ↕ SLOT_SIZE bytes
+//! ----------------------------------------
+//! ...
+//! ----------------------------------------
+//! E_(L-1)                ↕ SLOT_SIZE bytes
+//! ----------------------------------------
+//! Unallocated space
+//! ```
+//!
+//! The `SLOT_SIZE` constant depends on the item type. If the item
+//! type sets the `BoundedStorable::IS_FIXED_SIZE` flag, the
+//! `SLOT_SIZE` is equal to `BoundedStorable::MAX_SIZE`.  Otherwise,
+//! the `SLOT_SIZE` is `BoundedStorable::MAX_SIZE` plus the number of
+//! bytes required to represent integers up to
+//! `BoundedStorable::MAX_SIZE`.
+use crate::{
+    read_u32, read_u64, safe_write, write_u32, write_u64, Address, BoundedStorable, GrowFailed,
+    Memory,
+};
+use std::borrow::Borrow;
+use std::fmt;
+use std::marker::PhantomData;
+
+#[cfg(test)]
+mod tests;
+
+const MAGIC: [u8; 3] = *b"SVC"; // Short for "stable vector".
+const LAYOUT_VERSION: u8 = 1;
+/// The offset where the user data begins.
+const DATA_OFFSET: u64 = 64;
+/// The offset where the vector length resides.
+const LEN_OFFSET: u64 = 4;
+
+#[derive(Debug)]
+struct HeaderV1 {
+    magic: [u8; 3],
+    version: u8,
+    len: u64,
+    max_size: u32,
+    is_fixed_size: bool,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum InitError {
+    /// The memory already contains another data structure.
+    /// Use [Vec::new] to overwrite it.
+    BadMagic,
+    /// The current version of [Vec] does not support the of the
+    /// memory layout.
+    IncompatibleVersion {
+        last_supported_version: u8,
+        decoded_version: u8,
+    },
+    /// The vector type is not compatible with the current vector
+    /// layout: MAX_SIZE and/or IS_FIXED_SIZE differ from the original
+    /// initialization parameters.
+    IncompatibleElementType,
+    /// Failed to allocate memory for the vector.
+    OutOfMemory,
+}
+
+/// An implementation of growable arrays in stable memory.
+pub struct Vec<T: BoundedStorable, M: Memory> {
+    memory: M,
+    _marker: PhantomData<T>,
+}
+
+impl<T: BoundedStorable, M: Memory> Vec<T, M> {
+    /// Creates a new empty vector in the specified memory,
+    /// overwriting any data structures the memory might have
+    /// contained previously.
+    ///
+    /// Complexity: O(1)
+    pub fn new(memory: M) -> Result<Self, GrowFailed> {
+        let header = HeaderV1 {
+            magic: MAGIC,
+            version: LAYOUT_VERSION,
+            len: 0,
+            max_size: T::MAX_SIZE,
+            is_fixed_size: T::IS_FIXED_SIZE,
+        };
+        Self::write_header(&header, &memory)?;
+        Ok(Self {
+            memory,
+            _marker: PhantomData,
+        })
+    }
+
+    /// Initializes a vector in the specified memory.
+    ///
+    /// Complexity: O(1)
+    ///
+    /// PRECONDITION: the memory is either empty or contains a valid
+    /// stable vector.
+    pub fn init(memory: M) -> Result<Self, InitError> {
```

Comment:
Curious, do you think the `InitError` returned here and the `GrowFailed` elsewhere are recoverable at runtime? If not, then why return the error as opposed to using asserts?

Reply:
> why return the error as opposed to using asserts?

I see two reasons why typed errors are better in `new` and `init`:
1. Explicit errors make testing better.
2. The user can attach a custom error message with an `expect`. This ability is essential if you have multiple data structures of the same kind. If one of them blows up, you want to know which one it was. Data structure constructors don't have enough context.

Furthermore, I think `Storable` would benefit from error handling in `from_bytes` as well so that the data structure code could report the context where the error happened, something like "failed to decode entry at index 5: ...".

Reply:
I agree that creating stable structures is rare and it's ok to add extra verbosity there. Two follow-ups:

1. This PR also adds errors to the main API. The `push` method returns a `GrowFailed`, and I don't expect the use of the `push` API to be rare. The inconvenience there can add up, especially that, even if the user knows which specific call to `push` lead to growing the memory to fail, there's still nothing they can do about it.

2. Why is `set` an exception? The error in `set` is arguably more likely to occur than the error in `push`, and it's something the user can/should protect against.

I'd suggest making assertions in the `push` API that growing the memory succeeds so that it's more convenient and more consistent with `set`, but that's not a strong opinion, so feel free to move forward with whatever option appeals to you.

Reply:
> Why is set an exception? The error in set is arguably more likely to occur than the error in push, and it's something the user can/should protect against.

`set` can panic only due to a programming error when the user messes up the indices (just like regular vector assignment). `push` returns an error under regular operation when the memory fills up.
The `push` error is also recoverable: for example, and archive canister can signal to the main canister that it's full so it's time to create another archive. I can't imagine a case where `set` error would be recoverable.